### ISSUE TITLE
Diffusion coefficient

### ISSUE THREAD

**Issue by delcmo** [mfem/mfem#1031]
Hi,

I wrote a class that inherits from `Coefficient` to compute a diffusion coefficient that depends on the numerical solution itself. The diffusion coefficient in the diffusion operator is set to be calculated with this new class. I implemented `Eval` but it seems that it is only called at the initialization stage: the diffusion coefficient is not updated during each time step. Do I have to call a function to update the diffusion coefficient at each time step? Is there an example I can rely on to debug my case?

Thanks,

Marco

**Comment by mlstowell** [comment]
Hi, @delcmo,

I expect you pass your diffusion coefficient into a `DiffusionIntegrator` and this is then passed into a `BilinearForm` object.  If so, your coefficient will be accessed during the call to `BilinearForm::Assemble()`.  In order to recalculate the diffusion operator using your coefficient you'll need to do the following:
```
// Assuming your BilinearForm object is called 'a'
a.Update();   // to clear the previous diffusion operator
a.Assemble(); // to calculate the new diffusion operator
a.Finalize(); // to prepare the operator for use
```

Best wishes,
Mark

**Comment by delcmo** [comment]
Mark,

thanks for the help.

Marco

**Comment by v-dobrev** [comment]
Hi @delcmo,

I think example 16/16p may be useful for you to take a look at -- it solves a similar problem using a delayed diffusion coefficient update.

**Comment by delcmo** [comment]
Thanks for your reply. I was missing the Update() call.

I am having another issue that is related to the same topic. The diffusion coefficient is a function of the fluid temperature that is computed from the density, the momentum and the energy. I solve for the density, the momentum and the energy. I set the constructor as follows:

`DiffusionCoefficient::DiffusionCoefficient(const mfem::ParFiniteElementSpace &fes_,
                                           const mfem::ParFiniteElementSpace &dfes_,
                                           mfem::BlockVector &u_block_,
                                           const mfem::Array<int> offsets_)`

and then implement the following in `Eval`:

```
  mfem::ParGridFunction rho(&fes);
  rho.SetFromTrueDofs(u_block.GetBlock(0));
  double rho_lcl = rho.GetValue(T.ElementNo, ip);
```

The last line fails in parallel and I am not sure to understand why. The error message is the following: 

Assertion failed: (i >= 0 && i < size) is false:
 --> Row index 0 is out of range [0,-1)
 ... in function: void mfem::Table::GetRow(int, Array<int> &) const
 ... in file: /Users/mxd/ORNL-Research/MFEM/mfem/general/table.cpp:194

libc++abi.dylib: terminating with uncaught exception of type mfem::ErrorException: Assertion failed: (i >= 0 && i < size) is false:
 --> Row index 0 is out of range [0,-1)
 ... in function: void mfem::Table::GetRow(int, Array<int> &) const
 ... in file: /Users/mxd/ORNL-Research/MFEM/mfem/general/table.cpp:194

Since it works in serial, the processor is trying to access a value that is not available. `T.ElementNo` returns a value that is larger than the size of the vector.

What would be the correct syntax to ensure that `rho` is properly defined?

Thanks,

Marco

**Comment by mlstowell** [comment]
Hi, @delcmo ,

If your `rho` field is a Discontinuous Galerkin field then in parallel the `ParGridFunction` contains an auxiliary vector to store data from neighboring processors (@v-dobrev had to point this out to me only a few weeks ago).  To populate and update this auxiliary vector you need to call `ParGridFunction::ExchangeFaceNbrData()` after `ParGridFunction::SetFromTrueDofs()` and before `ParGridFunction::GetValue()`.

Best wishes,
Mark

**Comment by delcmo** [comment]
Hi Mark,

thanks again for your help. I forgot to mention that I am working with DG fields. 

It partially solved the problem. I no longer get the error message but after a few calls it just hangs there. I am not sure what is causing it. In example 16, the diffusion coefficient is computed using `GridFunction` object that is then converted to a `GridFunctionCoefficient` object. Is that the correct approach? 

Also, in my case I initialize the `ParGridFunction` storing the density from a BlockVector object. I could also use the `ParGridFunction sol` that is initialized from the BlockVector and passed to the ODE solver. What is the best strategy/practice?

Thanks,

Marco
